### PR TITLE
Implement MultiFrameImageStreamCompleter - handles scheduling for animated image frames

### PR DESCRIPTION
**Pull request by amirh**
This CL requires https://github.com/flutter/engine/commit/2920d61a5c1a64e87be424d7dd4bd7a6c4cb3acc so expect Travis to be complaining until we successfully roll the engine.

https://github.com/flutter/flutter/issues/204

### CODE REVIEW

File: packages/flutter/lib/src/services/image_stream.dart [hunk 1]
```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
     });
   }
 }
+
+/// Manages the decoding and scheduling of image frames.
+/// 
+/// This class deals with 2 types of frames:
+///  * image frames - image frames of an animated image.
```
Comment (Hixie) [inline packages/flutter/lib/src/services/image_stream.dart]: you'll need a blank link before the first bullet for this to render correctly

File: packages/flutter/lib/src/services/image_stream.dart [hunk 2]
```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
...
+  MultiFrameImageStreamCompleter({
+    @required Future<ui.Codec> codec,
+    @required double scale,
+    InformationCollector informationCollector
+  }) :
+    _informationCollector = informationCollector,
+    _scale = scale,
+    _framesEmitted = 0,
+    _timer = null
+  {
```
Comment (Hixie) [inline packages/flutter/lib/src/services/image_stream.dart]: nit: this open brace should be on the previous line

Comment (Hixie) [reply packages/flutter/lib/src/services/image_stream.dart]: I believe dartfmt may even get this right if you put a comma at the end of the last named argument

File: packages/flutter/lib/src/services/image_stream.dart [hunk 3]
```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
...
+  ui.FrameInfo _nextFrame;
+  // When the current was first shown.
+  Duration _shownTimestamp;
+  // The requested duration for the current frame;
+  Duration _frameDuration;
+  // How many frames have been emitted so far.
+  int _framesEmitted;
+  Timer _timer;
+
+  void _onCodecReady(ui.Codec codec){
```
Comment (Hixie) [inline packages/flutter/lib/src/services/image_stream.dart]: I think _handleCodecReady is probably a more conventional name for this
see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#naming-rules-for-typedefs-and-function-variables

File: packages/flutter/lib/src/services/image_stream.dart [hunk 4]
```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
...
+    print('codec ready');
```
Comment (Hixie) [inline packages/flutter/lib/src/services/image_stream.dart]: i suspect you didn't intend this :-)

File: packages/flutter/lib/src/services/image_stream.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/lib/src/services/image_stream.dart]: (same with the other _onFoo names below)

File: packages/flutter/lib/src/services/image_stream.dart [hunk 5]
```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
...
+      _shownTimestamp = timestamp;
+      _frameDuration = _nextFrame.duration;
+      _nextFrame = null;
+      int completedCycles = (_framesEmitted / _codec.frameCount).floor();
+      if (_codec.repetitionCount == -1 || completedCycles <= _codec.repetitionCount) {
+        _decodeNextFrameAndSchedule();
+      }
+      return;
+    }
+    _timer = new Timer(_frameDuration - (timestamp - _shownTimestamp),
```
Comment (Hixie) [inline packages/flutter/lib/src/services/image_stream.dart]: this should take into account the timeDilation

File: packages/flutter/lib/src/services/image_stream.dart [hunk 6]
```diff
@@ -274,3 +288,163 @@ class OneFrameImageStreamCompleter extends ImageStreamCompleter {
...
+
+  void _onAppFrame(Duration timestamp) {
+    if (!_hasActiveListeners)
+      return;
+    if (_isFirstFrame() || _hasFrameDurationPassed(timestamp)) {
+      _emitFrame(new ImageInfo(image: _nextFrame.image, scale: _scale));
+      _shownTimestamp = timestamp;
+      _frameDuration = _nextFrame.duration;
+      _nextFrame = null;
+      int completedCycles = (_framesEmitted / _codec.frameCount).floor();
```
Comment (Hixie) [inline packages/flutter/lib/src/services/image_stream.dart]: you can use the `~/` operator here i believe

**Comment by Hixie** [comment]
LGTM
It's hard for me to 100% see if this is correct without also seeing the corresponding ImageProvider changes that use this, but LGTM so far.

File: packages/flutter/lib/src/services/image_stream.dart (hunk 4 above)
Comment (amirh) [reply packages/flutter/lib/src/services/image_stream.dart]: Oops, I actually removed these before pushing just didn't save the file.... :flushed:

File: packages/flutter/lib/src/services/image_stream.dart (hunk 5 above)
Comment (amirh) [reply packages/flutter/lib/src/services/image_stream.dart]: Looks like the timestamps handed to the app frame callbacks are already dilated... so we don't need to do anything here...

Comment (amirh) [reply packages/flutter/lib/src/services/image_stream.dart]: Ohh you're right, this timer is not affected by dilation, multiplied by time dilation.

**Pull request merged**